### DISCUSSION TITLE
Drivers no longer listed in finishing order in Laps

### DISCUSSION THREAD

**Discussion by Casper-Guo** [theOehrly/Fast-F1#373] [General]
In the past, `session.Laps` will list the drivers by the official classification order. I'm not sure if this was ever guaranteed but I haven't encountered any violation of this ordering in my usage.

This is no longer the case. Drivers are now listed in a fixed but seemingly random order.

# Code sample:
```
import fastf1 as f

hungary = f.get_session(2022, "Hungary", 'R')
hungary.load()

print(hungary.laps["Driver"].unique())

azerbaijan = f.get_session(2023, "Azerbaijan", 'R')
azerbaijan.load()

print(azerbaijan.laps["Driver"].unique())

silverstone = f.get_session(2022, "Silverstone", 'R')
silverstone.load(telemetry=False, weather=False, messages=False)

print(silverstone.laps["Driver"].unique())
```

# Output
```
['VER' 'GAS' 'PER' 'ALO' 'LEC' 'STR' 'MAG' 'TSU' 'ALB' 'ZHO' 'RIC' 'OCO'
 'NOR' 'HAM' 'MSC' 'VET' 'SAI' 'LAT' 'RUS' 'BOT']
['VER' 'GAS' 'PER' 'ALO' 'LEC' 'STR' 'SAR' 'MAG' 'DEV' 'TSU' 'ALB' 'ZHO'
 'HUL' 'OCO' 'NOR' 'HAM' 'SAI' 'RUS' 'BOT' 'PIA']
['VER' 'GAS' 'PER' 'ALO' 'LEC' 'STR' 'MAG' 'TSU' 'ALB' 'ZHO' 'RIC' 'OCO'
 'NOR' 'HAM' 'MSC' 'VET' 'SAI' 'LAT' 'RUS' 'BOT']
```

# Relevant
Pandas series' unique function returns items in order of appearance.

> Uniques are returned in order of appearance. Hash table-based unique, therefore does NOT sort.

Just curious if this is an intentional change or is caused by the upstream API.

**Comment by theOehrly** [comment]
I need to look into this some more, maybe tomorrow. I don't think this was ever an intentional feature. At least I don't remeber it. 

``Session.drivers`` should still return drivers in the sorted order because the data is taken from ``Seession.results`` when results are available.

**Comment by Casper-Guo** [comment]
Sounds good, just refactored the strategy example (#372) to use `Session.drivers` for the time being.

**Comment by theOehrly** [comment]
OK, so this is caused by a change in FastF1. But it was never intended that the laps are sorted by finishing order. Currently, some sessions will be ordered by finishing order and some won't be. 
I don't really want to start guaranteeing that laps are sorted by finishing order. Instead, it might be more reasonable to ensure a fixed order by either sorting them by driver number or abbreviation. What do you think?

**Comment by Casper-Guo** [comment]
My personal order of preference:

1. Sorted by finishing order/official classification. This is the most intuitive order for talking about, for example, the "first 10" drivers in a session. This will also make finding the point finishers or qualification eliminations trivial and without needing to use `session.result` or similar.
2. Sorted by another reasonable attribute, driver number or abbreviation are both good options.
3. Some other fixed ordering.
4. No fix ordering (which is apparently the current state 😨 )

**Reply by theOehrly** [reply]
My problem with 1) is, what order do we use for free practice? Option 2) then? Also, I need to test the performance cost of guaranteeing that the data is ordered by finishing order. 

Furthermore, the session results depend on Ergast which updates slower than the F1 API. Therefore, I won't be able to reliably sort by correct finishing order until a few hours after where data starts being available. I can only sort by calculated finishing order, which does not include time penalties, for example. Therefore, the order might change again, once data from Ergast becomes available. (The same actually applies to `Session.drivers` which is why I am not sure that I want to guarantee any particular order there. Simply because I can't.) 

The only truly reliable way to get the finishing order is by using `SessionResults`. If data is available, that is the final order. If no data is available, you'll have to wait until Ergast is updated.

It is possible that I might start to run my own server backend infrastructure at some point to work around issues like this (I kind of already do, because of the event schedule). But the setup work and maintenance required for this is difficult to justify. 

Regarding 4), the `Laps` DataFrame was only ever intended to be a collection of Laps and nothing more.

**Reply by Casper-Guo** [reply]
F1 also provides official classifications for free practices but I am guessing you mean it is not provided by Ergast. Option 2) is a good middle ground in that case.
